### PR TITLE
DNM: bluestore: FreeBSD make `ceph-osd mkfs` use AIO.

### PR DESCRIPTION
**Pull request by wjwithagen**
- build the correct aio blocks
 - make kevent return completed events and pass
   the original aio struct as priv pointer

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/os/bluestore/BlockDevice.cc [hunk 1]
```diff
@@ -84,16 +85,22 @@ void IOContext::release_running_aios()
...
 BlockDevice *BlockDevice::create(CephContext* cct, const string& path,
-				 aio_callback_t cb, void *cbpriv, aio_callback_t d_cb, void *d_cbpriv)
+                                 aio_callback_t cb, void *cbpriv, 
+                                 aio_callback_t d_cb, void *d_cbpriv)
 {
   string type = "kernel";
   char buf[PATH_MAX + 1];
   int r = ::readlink(path.c_str(), buf, sizeof(buf) - 1);
+
+  dout(1) << __func__ << " path " << path << " type " << type << dendl;
```
Comment (tchaikov) [inline src/os/bluestore/BlockDevice.cc]: why is this relevant to "FreeBSD make `ceph-osd mkfs` use AIO" ?

Comment (wjwithagen) [reply src/os/bluestore/BlockDevice.cc]: @tchaikov 
Because it is good to know what devices bluestore is attaching too, certainly during debuging, but also durign regular use.

**Comment by tchaikov** [comment]
@wjwithagen ping

**Comment by wjwithagen** [comment]
@tchaikov 
Yes I know, there is a lot of work hanging out to dry?
But lately I got caught in a lot of work for the dayjob... 

And before anything I still have to fix the `msg_type sa_len` stuff for FreeBSD.
Which I got almost nailed, except for a few odd bytes in the `addr_vec_t` which do not seem to matter in most cases, but they let a `ceph-dencoder decode encode decode` crash.
And these are not real easy to figure out.

So yes, this one needs to stay for the time being. So I'll probably just rebase.

**Comment by neha-ojha** [comment]
needs rebase

@tchaikov ping

**Comment by wjwithagen** [comment]
@neha-ojha
I'm in the process of getting all my jenkins testers to generate testruns without problems.
After that I'll return to fixing the outstanding PRs, including rebasing.
But there has been some work on the AIO innards, so I expect it 'll be more that just fixing some shifted lines....

File: src/os/bluestore/BlockDevice.cc (hunk 1 above)
Comment (tchaikov) [reply src/os/bluestore/BlockDevice.cc]: @wjwithagen i suggest split this change into smaller commits, to avoid mixing the changes changing existing logging messages with the change introducing POSIX aio support.

File: src/os/bluestore/BlockDevice.cc [hunk 2]
```diff
@@ -84,16 +85,22 @@ void IOContext::release_running_aios()
...
+
   if (r >= 0) {
     buf[r] = '\0';
+#if defined(HAVE_SPDK)
```
Comment (tchaikov) [inline src/os/bluestore/BlockDevice.cc]: might want to move this change into a different commit.

Comment (wjwithagen) [reply src/os/bluestore/BlockDevice.cc]: @tchaikov 
Here plays my git ignorance again.... :-(
How do I split the current 1 commit in 3 parts?
Suggestions

Comment (tchaikov) [reply src/os/bluestore/BlockDevice.cc]: i'd suggest
```
git reset HEAD^
git add -p
```

**Comment by wjwithagen** [comment]
@tchaikov @liewegas 
This allows me to also run `unittest_{bluefs,bluestore_types}` on FreeBSD, which gives me a better feeling that the current patchset is doing "(tm) the right thing" 
Are there more basic test (other than a full cluster) that can verify this?

File: src/os/bluestore/KernelDevice.cc [hunk 3]
```diff
@@ -201,7 +201,7 @@ int KernelDevice::open(const string& p)
 	  << " block_size " << block_size
 	  << " (" << byte_u_t(block_size) << ")"
 	  << " " << (rotational ? "rotational" : "non-rotational")
-      << " discard " << (support_discard ? "supported" : "not supported")
+       << " discard " << (support_discard ? "supported" : "not supported")
```
Comment (tchaikov) [inline src/os/bluestore/KernelDevice.cc]: could you drop this change?

File: src/os/bluestore/aio.cc [hunk 4]
```diff
@@ -1,9 +1,19 @@
 // -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
 // vim: ts=8 sw=2 smarttab
 
+#include "acconfig.h"
+#include "include/compat.h"
```
Comment (tchaikov) [inline src/os/bluestore/aio.cc]: might want to move these two includes after `#include <algorithm>`.

File: src/os/bluestore/aio.cc [hunk 5]
```diff
@@ -15,6 +25,50 @@ std::ostream& operator<<(std::ostream& os, const aio_t& aio)
...
+}
+#endif
+
+int aio_queue_t::init(int f) {
+    ceph_assert(ctx == 0);
+#if defined(HAVE_LIBAIO)
+    int r = io_setup(max_iodepth, &ctx);
+    if (r < 0) {
+      if (ctx) {
+      io_destroy(ctx);
```
Comment (tchaikov) [inline src/os/bluestore/aio.cc]: wrong indent.

File: src/os/bluestore/aio.cc [hunk 6]
```diff
@@ -71,54 +167,133 @@ int aio_queue_t::submit_batch(aio_iter begin, aio_iter end,
...
-	}
-      }
-      free(paio[i]->aio.aiocbp);
-    }
-#endif
   }
   return r;
 }
+#endif
+#if defined(HAVE_POSIXAIO)
```
Comment (tchaikov) [inline src/os/bluestore/aio.cc]:
```c++
#elif defined(...)
```

File: src/os/bluestore/ceph_aio.h [hunk 7]
```diff
@@ -144,5 +137,11 @@ struct aio_queue_t {
 
   int submit_batch(aio_iter begin, aio_iter end, uint16_t aios_size, 
 		   void *priv, int *retries);
+#if defined(HAVE_LIBAIO)
   int get_next_completed(int timeout_ms, aio_t **paio, int max);
+#endif
```
Comment (tchaikov) [inline src/os/bluestore/ceph_aio.h]: `#elif defined`

File: src/os/bluestore/ceph_aio.h [hunk 8]
```diff
@@ -22,13 +22,13 @@ struct aio_t {
...
   struct iocb iocb{};  // must be first element; see shenanigans in aio_queue_t
 #elif defined(HAVE_POSIXAIO)
-  //  static long aio_listio_max = -1;
   union {
     struct aiocb aiocb;
     struct aiocb *aiocbp;
   } aio;
   int n_aiocb;
 #endif
+  CephContext *cct;
```
Comment (tchaikov) [inline src/os/bluestore/ceph_aio.h]: i am not sure if we want to add yet another 8-byte member variable in `aio_t`. probably we can enable it only in debug build?

@liewegas what do you think?

File: src/os/bluestore/aio.cc [hunk 9]
```diff
@@ -71,54 +167,133 @@ int aio_queue_t::submit_batch(aio_iter begin, aio_iter end,
...
+            << dendl;
+        return rev;
+      } else {
+        dout(25) << __func__ << " process " << rev << " events"
+            << dendl;
+        // Process all the events posted
+        // Emulate the return value of pwritev. I can't find any documentation
+        // for what the value of io_event.res is supposed to be. I'm going to
+        // assume that it's just like pwritev/preadv/pwrite/pread.
+        for(int i=0; i < rev; i++) {
```
Comment (tchaikov) [inline src/os/bluestore/aio.cc]: add spaces after `for` and around `=`.

File: src/os/bluestore/aio.cc [hunk 10]
```diff
@@ -39,18 +98,55 @@ int aio_queue_t::submit_batch(aio_iter begin, aio_iter end,
...
+      piocb[done]->aio.aiocbp[0].aio_sigevent.sigev_value.sival_ptr =
+                                                    (void*)&(piocb[done]->aio);
+      switch (piocb[done]->aio.aiocbp[0].aio_lio_opcode) {
+        case LIO_WRITE:
+          r = aio_write(&(piocb[done]->aio.aiocbp[0]));
+          break;
+        case LIO_READ:
+          r = aio_read(&(piocb[done]->aio.aiocbp[0]));
+          break;
+        default:
```
Comment (tchaikov) [inline src/os/bluestore/aio.cc]: might want to handle `LIO_NOP` as well?

File: src/os/bluestore/aio.cc [hunk 11]
```diff
@@ -15,6 +25,50 @@ std::ostream& operator<<(std::ostream& os, const aio_t& aio)
...
+      }
+    }
+    return r;
+#elif defined(HAVE_POSIXAIO)
+    ctx = kqueue();
+    if (ctx < 0)
+      return -errno;
+    else
+      return 0;
+    fd = f;
```
Comment (tchaikov) [inline src/os/bluestore/aio.cc]: is this dead code?

File: src/os/bluestore/aio.cc [hunk 12]
```diff
@@ -15,6 +25,50 @@ std::ostream& operator<<(std::ostream& os, const aio_t& aio)
   return os;
 }
 
+#if defined(__FreeBSD__)
```
Comment (tchaikov) [inline src/os/bluestore/aio.cc]: might want to use `HAVE_POSIXAIO` instead, to be more consistent this way.

File: src/os/bluestore/aio.cc [hunk 13]
```diff
@@ -15,6 +25,50 @@ std::ostream& operator<<(std::ostream& os, const aio_t& aio)
...
+  os << ", priv.error = " << cb->_aiocb_private.error;
+  os << ", priv.sigevent.notify = " << cb->aio_sigevent.sigev_notify;
+  os << ", priv.sigevent.notify_kqueue = " << cb->aio_sigevent.sigev_notify_kqueue;
+  os << ", priv.sigevent.kevent_flags = " << cb->aio_sigevent.sigev_notify_kevent_flags;
+  os << ")";
+  return os;
+}
+#endif
+
+int aio_queue_t::init(int f) {
```
Comment (tchaikov) [inline src/os/bluestore/aio.cc]: what `f` is for?

Comment (tchaikov) [inline src/os/bluestore/aio.cc]: please split this change into functional changes and non-functional changes to help reviewer to understand your changes. in this case, you could have two changes: one to move the init() to .cc file. another to improve it. it's difficult to me to track the changes.

File: src/os/bluestore/aio.cc (hunk 4 above)
Comment (wjwithagen) [reply src/os/bluestore/aio.cc]: @tchaikov 
I will, but can you explain why?

Because I advise against including (boost) C++ files before `include/compat.h` since it will interfere with the definition of ENODATA.

File: src/os/bluestore/aio.cc (hunk 13 above)
Comment (wjwithagen) [reply src/os/bluestore/aio.cc]: @tchaikov 
Perhaps better leaving the complete body in the headerfile then?

File: src/os/bluestore/aio.cc (hunk 4 above)
Comment (tchaikov) [reply src/os/bluestore/aio.cc]: see https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

yeah, but i think `<algorithm>` does not fall into that bucket.

File: src/os/bluestore/aio.cc (hunk 10 above)
Comment (wjwithagen) [reply src/os/bluestore/aio.cc]: @tchaikov 
I do not see any code issue a `LIO_NOP`, so you think it is really needed?

Now it is caugth by the default. I've initialize `r = 0` on every itteration, so it is counted as a successful instruction.

File: src/os/bluestore/ceph_aio.h (hunk 8 above)
Comment (wjwithagen) [reply src/os/bluestore/ceph_aio.h]: @tchaikov 
I want the context there so I can do logging from the aio-functions.
Doing so in only debug builds is workable for the time being. 
Not sure what to do when we need to diagnose non-debug external runs.

File: src/os/bluestore/aio.cc (hunk 10 above)
Comment (tchaikov) [reply src/os/bluestore/aio.cc]: just for the sake of completeness. i think we can just list it as another 
```c++
case LIO_NOP:
  break;
```
instead of using `default` to catch it.

File: src/os/bluestore/aio.cc (hunk 13 above)
Comment (wjwithagen) [reply src/os/bluestore/aio.cc]: @tchaikov 
Moved the init() function back into the headerfile.

Are there more things to do before you can take this PR for merging?

File: src/os/bluestore/aio.cc [hunk 14]
```diff
@@ -71,54 +145,132 @@ int aio_queue_t::submit_batch(aio_iter begin, aio_iter end,
...
+  struct kevent events[max];
+  struct kevent ev;
+  dout(21) << __func__ << " kqueue registration"
+      << " on fd = " << fd
+      << " , with kqueue handle ctx = " << ctx
+      << dendl;
+
+  EV_SET(&ev, fd, 0, EV_ADD | EV_CLEAR, 0, 0, 0);
+  int rev = kevent(ctx, &ev, 1, NULL, 0, 0 );
+  if (rev < 0) {
```
Comment (tchaikov) [inline src/os/bluestore/aio.cc]: should also check for `ev.flags & EV_ERROR`.

File: src/os/bluestore/aio.cc [hunk 15]
```diff
@@ -71,54 +145,132 @@ int aio_queue_t::submit_batch(aio_iter begin, aio_iter end,
...
+  };
+
+  struct kevent events[max];
+  struct kevent ev;
+  dout(21) << __func__ << " kqueue registration"
+      << " on fd = " << fd
+      << " , with kqueue handle ctx = " << ctx
+      << dendl;
+
+  EV_SET(&ev, fd, 0, EV_ADD | EV_CLEAR, 0, 0, 0);
```
Comment (tchaikov) [inline src/os/bluestore/aio.cc]: i am not sure why you are changing
```c++
r = kevent(ctx, NULL, 0, events, max, &t);
```

to 
```c++
int rev = kevent(ctx, &ev, 1, NULL, 0, 0 );
```
?

File: src/os/bluestore/aio.cc [hunk 16]
```diff
@@ -71,54 +145,132 @@ int aio_queue_t::submit_batch(aio_iter begin, aio_iter end,
...
+  int rev = kevent(ctx, &ev, 1, NULL, 0, 0 );
+  if (rev < 0) {
+    dout(1) << __func__ << " kqueue registration error "
+        << " kevent return(" << errno << ") "
+        << cpp_strerror(errno)
+        << dendl;
+    return -errno;
+  }
+  do {
```
Comment (tchaikov) [inline src/os/bluestore/aio.cc]: what is this loop for?

File: src/os/bluestore/ceph_aio.h [hunk 17]
```diff
@@ -99,26 +102,34 @@ struct aio_queue_t {
...
   int init() {
     ceph_assert(ctx == 0);
 #if defined(HAVE_LIBAIO)
     int r = io_setup(max_iodepth, &ctx);
     if (r < 0) {
       if (ctx) {
-	io_destroy(ctx);
-	ctx = 0;
+        io_destroy(ctx);
+      ctx = 0;
```
[the remark this answers is not in the document]
Comment (wjwithagen) [reply src/os/bluestore/ceph_aio.h]: @tchaikov 
This needs to be indented, but I fixed the misalignment.

File: src/os/bluestore/aio.cc (hunk 16 above)
Comment (wjwithagen) [reply src/os/bluestore/aio.cc]: @tchaikov 
Since it is async IO, the event-queue can hold multiple completion events.
And we just process all outstanding IO's completed that are available at this time, and return that in 1 completion structure.

File: src/os/bluestore/aio.cc (hunk 15 above)
Comment (wjwithagen) [reply src/os/bluestore/aio.cc]: @tchaikov 
Good question, and I'll have to dig this up.
There is not a lot of documentation and examples for `kqueue()` code, and I think I got this from IRC with another FreeBSD programmer.
But the first call is to set the filter, and check if that works ok.
The second call in the `do() while` is to fetch events from the queue

File: src/os/bluestore/aio.cc (hunk 14 above)
Comment (wjwithagen) [reply src/os/bluestore/aio.cc]: We are registering the event-filter. Resulting events are normally returned in the last three parameters, which are all unavailable because the size of the event-array is zero. 
So errors are returned by function-result and errno.

Later on you have the opposite, the filter is not changed (ev=NULL), and we are only asking for events to be delivered. (using the previously set filter)

File: src/os/bluestore/ceph_aio.h (hunk 8 above)
Comment (wjwithagen) [reply src/os/bluestore/ceph_aio.h]: @tchaikov 
I'll remove the logging.
Would it be valid to write the errors to stderr?? That way there is some indication that things are not oke...

**Comment by wjwithagen** [comment]
Keep it, perhaps I get around to fixing/rebasing this during the holidays.

**Comment by wjwithagen** [comment]
Revive stale

File: src/os/bluestore/ceph_aio.h (hunk 8 above)
Comment (tchaikov) [reply src/os/bluestore/ceph_aio.h:31]: @wjwithagen  if the daemon is daemonized, stderr will be closed. i am afraid we cannot rely on `stderr` for logging.

File: src/os/bluestore/ceph_aio.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/os/bluestore/ceph_aio.h:145]: we use tabstop = 8. i cannot see anything wrong with the alignment or indent.

File: src/os/bluestore/BlockDevice.cc [hunk 18]
```diff
@@ -93,9 +93,11 @@ BlockDevice *BlockDevice::create(CephContext* cct, const string& path,
   int r = ::readlink(path.c_str(), buf, sizeof(buf) - 1);
   if (r >= 0) {
     buf[r] = '\0';
+#if defined(HAVE_SPDK)
```
Comment (tchaikov) [inline src/os/bluestore/BlockDevice.cc:101]: "Exclude nvme if no SPDK": 
could you prefix the title of your commit message with the subcomponent your are changing ? see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#3-describe-your-changes. in this case, it'd be "os/bluestore:".

Comment (wjwithagen) [reply src/os/bluestore/BlockDevice.cc:101]: @tchaikov 
Thanx, for the review, and I'll fix your request.
I've set the PR to DNM due to other issues I'm having.
ATM I'm tracking down a weird error I have where I have the decode template for  `_u32` return a value equal to the first 4 chars of the string following it.
And that is hampered by the the recent upgrade to Clang10, so I'm battling at a few front to get things going.

**Comment by wjwithagen** [comment]
Unstale

**Comment by wjwithagen** [comment]
unstale

**Comment by wjwithagen** [comment]
unstale

**Comment by wjwithagen** [comment]
keep
